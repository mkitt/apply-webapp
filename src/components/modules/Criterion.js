// @flow
import React from 'react'
import glamorous from 'glamorous'
import { memoizeWith, identity, length } from 'ramda'
import { pure } from 'recompose'
import { COLORS } from '../../constants/theme'
import { ChevronXIcon } from '../elements/Icons'
import type { Criterion } from '../../types/Criterion'

type Props = {
  criterion: Criterion,
  isOpen: boolean,
}

const len = memoizeWith(identity, list => length(list) - 1)

export default pure(({ criterion, isOpen }: Props) => (
  <Details open={isOpen}>
    <Summary>
      <span>{criterion.label}</span>
      <ChevronXIcon />
    </Summary>
    <Content>
      <Questions>
        {criterion.questions.map(question => (
          <Question key={question}>
            {question}
          </Question>
        ))}
      </Questions>
      <OptionDl>
        {criterion.options.map((option, index) => (
          <OptionItem key={option}>
            <OptionDt>{len(criterion.options) - index}</OptionDt>
            <OptionDd>{option}</OptionDd>
          </OptionItem>
        ))}
      </OptionDl>
    </Content>
  </Details>
))

// -------------------------------------

const Details = glamorous.details({
})

const Summary = pure(glamorous.summary({
  position: 'relative',
  minHeight: 48,
  fontSize: 28,
  fontWeight: 200,
  color: COLORS.GREY_8,
  borderBottom: '1px solid',
  outline: 0,
  cursor: 'pointer',
  '&::-webkit-details-marker': {
    display: 'none',
  },
  '&:first-of-type': {
    listStyleType: 'none',
  },
  'details[open] > &': {
    color: COLORS.GREY_4,
  },
  '& > svg': {
    position: 'absolute',
    top: 8,
    right: 0,
  },
  // Yo! Using a `not` filter here
  'details:not([open]) > & > svg': {
    top: 16,
  },
  // Yo! Also using a `not` filter here
  'details:not([open]) > & > svg g:nth-child(2)': {
    opacity: 0,
  },
}))

const Content = glamorous.div({
  paddingTop: 44,
  paddingBottom: 48,
})

const Questions = glamorous.ol({
  listStyle: 'none',
  margin: 0,
  paddingTop: 0,
  paddingRight: 0,
  paddingBottom: 38,
  paddingLeft: 48,
  color: COLORS.GREY_4,
})

const Question = glamorous.li({
  fontSize: 16,
  margin: 0,
  marginBottom: 24,
})

const OptionDl = glamorous.dl({
  display: 'flex',
  justifyContent: 'space-between',
  alignItems: 'flex-start',
  minHeight: 160,
  margin: 0,
  marginTop: 14,
  paddingTop: 32,
  paddingRight: 48,
  paddingBottom: 42,
  paddingLeft: 48,
  color: COLORS.GREY_4,
  backgroundColor: COLORS.GREY_E,
})

const OptionItem = glamorous.div({
  paddingRight: 32,
})

const OptionDt = glamorous.dt({
  marginBottom: 14,
  fontSize: 28,
  fontWeight: 200,
})

const OptionDd = glamorous.dd({
  marginLeft: 0,
})
