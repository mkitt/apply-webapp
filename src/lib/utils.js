// @flow
import { __, addIndex, always, and, complement, either, gt, isEmpty, isNil, map } from 'ramda'
import { withProps } from 'recompose'

// -------------------------------------
// Environments
export const isDev = always(process.env.NODE_ENV === 'development')

// -------------------------------------
// Types
export const isNotNil = complement(isNil)
export const isNotEmpty = complement(isEmpty)
export const isEmptyOrNil = either(isNil, isEmpty)
export const isNotEmptyAndNotNil = and(isNotNil, isNotEmpty)

// -------------------------------------
// Relation
export const gt0 = gt(__, 0)

// -------------------------------------
// Lists
export const nool = always([])()
export const mapIndexed = addIndex(map)

// -------------------------------------
// Objects
export const noob = always({})()

// -------------------------------------
// Functions
export const noop = () => {}

export const log = withProps(props => (
  // eslint-disable-next-line no-console
  console.log(props.displayName || 'üëç', props) && props
))

export const warn = withProps(props => (
  // eslint-disable-next-line no-console
  console.warn(props.displayName || '‚ö†Ô∏è', props) && props
))
