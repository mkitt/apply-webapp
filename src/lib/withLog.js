import { withProps } from 'recompose'

export default withProps(props => (
  // eslint-disable-next-line
  console.log(props.logName || props.displayName || '👍', props) && props
))
