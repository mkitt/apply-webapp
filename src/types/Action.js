// @flow
import { camelize } from 'humps'
import type { Applicant } from './Applicant'

export type Action = {
  applicants: Array<Applicant>,
  id: ID,
  label: string,
  message: string,
  name: string,
}

// -------------------------------------
// For testing content
export const stub = (props: Action) => {
  const label = (props != null && props.label) || 'Mismatch'
  const camel = camelize(label)
  return {
    id: camel,
    label,
    message: `${label} message`,
    name: camel,
    ...props,
  }
}

export default undefined
