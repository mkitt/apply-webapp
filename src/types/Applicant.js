// @flow
import { camelize } from 'humps'
import type { Action } from './Action'
import type { Application } from './Application'

export type Applicant = {
  action: ?Action,
  applications: Array<Application>,
  birthdate: ?string,
  createdAt: DateTime,
  email: string,
  firstName: string,
  github: string,
  id: ID,
  lastName: string,
  referredBy: ?string,
  resume: ?string,
  startDate: ?DateTime,
  status: ?string,
  updatedAt: DateTime,
}

// -------------------------------------
// For testing content
export const stub = (props: Applicant) => {
  const firstName = (props != null && props.firstName) || 'Kenny'
  const lastName = (props != null && props.lastName) || 'Bania'
  const camel = camelize(`${firstName} ${lastName}`)
  const applicant = {
    id: camel,
    action: null,
    applications: [],
    birthdate: '06/06/2006',
    createdAt: '2017-10-31T12:12:12.000Z',
    email: 'kenny.bania@example.com',
    firstName: 'Kenny',
    github: camel,
    lastName,
    referredBy: 'Simon Taranto',
    resume: 'https://www.resume.com',
    startDate: null,
    status: 'pending',
    updatedAt: '2017-12-24T24:24:24.000Z',
  }
  return {
    ...applicant,
    ...props,
  }
}

export default undefined
