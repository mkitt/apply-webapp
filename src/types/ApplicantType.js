// @flow
export type ApplicantType = {
  applyAction: string,
  applyStatus: string,
  birthdate: string,
  createdAt: string,
  email: string,
  firstName: string,
  github: string,
  id: string,
  lastName: string,
  referredBy: string,
  resume: string,
  scoreLogicEvaluation: number,
  scoreOnlineLogicTest: number,
  scoreValuesEvaluation: number,
  startDate: string,
  updatedAt: string,
}
