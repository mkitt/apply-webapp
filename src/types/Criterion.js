// @flow
import { camelize } from 'humps'
import type { Evaluation } from './Evaluation'

export type Criterion = {
  id: ID,
  createdAt: DateTime,
  evaluation: ?Evaluation,
  label: string,
  name: string,
  options: Array<string>,
  questions: Array<string>,
  updatedAt: DateTime,
}

// -------------------------------------
// For testing content
export const stub = (props: Criterion) => {
  const label = (props != null && props.label) || 'Problem Solving'
  const camel = camelize(label)
  return {
    id: camel,
    label,
    name: camel,
    options: [
      `${label} Option 3`,
      `${label} Option 2`,
      `${label} Option 1`,
    ],
    questions: [
      `${label} Question 1`,
      `${label} Question 2`,
      `${label} Question 3`,
    ],
    ...props,
  }
}

export default undefined
