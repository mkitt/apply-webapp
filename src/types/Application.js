// @flow
import { both, compose, either, prop, propSatisfies, replace, test } from 'ramda'
import { capitalize, gt0 } from '../lib/utils'
import type { Applicant } from './Applicant'
import type { Evaluator } from './Evaluator'
import type { Evaluation } from './Evaluation'

export type Application = {
  id: ID,
  applicant: Applicant,
  complete: boolean,
  evaluations: Array<Evaluation>,
  evaluators: ?Array<Evaluator>,
  resume: ?string,
  scoreLogicEvaluation: ?Int,
  scoreOnlineLogicTest: ?Int,
  scoreValuesEvaluation: ?Int,
  status: string,
  // derived...
  isStatusMutatable: boolean,
  statusLabel: string,
}

// -------------------------------------
// Derived helpers

/*
export type StatusType = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14
export const STATUS = Object.freeze({
  pending: 0,
  needs_initial_evaluation_scores: 1,
  needs_rejected_at_initial_evaluation_notification: 2,
  rejected_at_initial_evaluation: 3,
  needs_to_schedule_interview: 4,
  needs_interview_scores: 5,
  needs_rejected_at_interview_notification: 6,
  rejected_at_interview: 7,
  needs_logic_evaluation_scores: 8,
  needs_rejected_at_logic_evaluation_notification: 9,
  rejected_at_logic_evaluation: 10,
  needs_invitation: 11,
  needs_invitation_response: 12,
  declined_invitation: 13,
  accepted_invitation: 14,
})
*/

export const needsSendStatus = compose(
  test(/^needs_rej|needs_invitation$/),
  prop('status'),
)

export const needsSubmitEvaluation = compose(
  test(/_evaluation_scores$|^needs_interview/),
  prop('status'),
)

export const deriveIsStatusMutatable = either(
  propSatisfies(test(/^needs_rej|needs_invitation$|_evaluation_scores$/), 'status'),
  both(
    propSatisfies(test(/^needs_interview/), 'status'),
    propSatisfies(gt0, 'scoreOnlineLogicTest'),
  ),
)

export const deriveStatusLabel = compose(
  capitalize,
  replace(/_/gi, ' '),
  prop('status'),
)

// -------------------------------------
// For testing content
export const stub = (props: Application) => {
  const application = {
    id: '1',
    complete: false,
    evaluations: [],
    evaluators: null,
    isStatusMutatable: false,
    resume: 'uploads/asset/application-resume.pdf',
    scoreLogicEvaluation: 20,
    scoreOnlineLogicTest: 40,
    scoreValuesEvaluation: 20,
    showScores: true,
    status: 'pending',
    statusLabel: 'Pending',
  }
  return {
    ...application,
    ...props,
  }
}

export default undefined
