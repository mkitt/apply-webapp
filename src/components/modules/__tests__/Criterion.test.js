import React from 'react'
import renderer from 'react-test-renderer'
import Criterion from '../Criterion'
import { noop } from '../../../lib/utils'
import { stub } from '../../../types/Criterion'

describe('<Criterion />', () => {
  it('shows the correct snapshot tree for the <Criterion> module', () => {
    const component = (
      <Criterion criterion={stub()} handleChange={noop} />
    )
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
