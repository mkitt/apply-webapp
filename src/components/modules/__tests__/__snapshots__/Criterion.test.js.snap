// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Criterion /> shows the correct snapshot tree for the <Criterion> module 1`] = `
.glamor-2,
[data-glamor-2] {
  position: relative;
  min-height: 48px;
  font-size: 28px;
  font-weight: 200;
  color: #808080;
  border-bottom: 1px solid;
  outline: 0;
  cursor: pointer;
}

.glamor-2::-webkit-details-marker,
[data-glamor-2]::-webkit-details-marker {
  display: none;
}

.glamor-2:first-of-type,
[data-glamor-2]:first-of-type {
  list-style-type: none;
}

.glamor-2 > svg,
[data-glamor-2] > svg {
  position: absolute;
  top: 8px;
  right: 0;
}

.glamor-0,
[data-glamor-0] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: space-between;
  padding-right: 48px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
}

.glamor-0 > span + span,
[data-glamor-0] > span + span {
  color: #35c5fc;
}

.glamor-22,
[data-glamor-22] {
  position: relative;
  padding-top: 44px;
  padding-bottom: 48px;
}

.glamor-7,
[data-glamor-7] {
  width: 256px;
}

.glamor-6,
[data-glamor-6] {
  position: relative;
}

.glamor-3,
[data-glamor-3] {
  -webkit-appearance: button;
  -webkit-border-radius: 0;
  -webkit-box-shadow: none;
  -webkit-padding-end: 0;
  -webkit-padding-start: 0;
  -webkit-user-select: none;
  width: 100%;
  height: 48px;
  margin: 0;
  padding: 0 32px;
  overflow: hidden;
  font-size: 14px;
  font-family: inherit;
  text-transform: uppercase;
  color: #fff;
  background-color: #4d4d4d;
  background-image: none;
  background-position: initial;
  background-repeat: no-repeat;
  border-radius: 0;
  border: 0;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.glamor-3:focus,
[data-glamor-3]:focus {
  border-color: #999;
  outline: 0;
}

.glamor-5,
[data-glamor-5] {
  position: absolute;
  top: calc(50% - 8px);
  right: 32px;
  color: #fff;
  pointer-events: none;
}

.glamor-11,
[data-glamor-11] {
  list-style: none;
  margin: 0;
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 38px;
  padding-left: 48px;
  color: #4d4d4d;
}

.glamor-8,
[data-glamor-8] {
  font-size: 16px;
  margin: 0;
  margin-bottom: 24px;
}

.glamor-21,
[data-glamor-21] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: space-between;
  align-items: flex-start;
  min-height: 160px;
  margin: 0;
  margin-top: 14px;
  padding-top: 32px;
  padding-right: 48px;
  padding-bottom: 42px;
  padding-left: 48px;
  color: #4d4d4d;
  background-color: #e5e5e5;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -webkit-box-align: start;
  -webkit-align-items: flex-start;
}

.glamor-14,
[data-glamor-14] {
  padding-right: 32px;
}

.glamor-12,
[data-glamor-12] {
  margin-bottom: 14px;
  font-size: 28px;
  font-weight: 200;
}

.glamor-13,
[data-glamor-13] {
  margin-left: 0;
}

@media (min-width: 48em) {
  .glamor-7,
  [data-glamor-7] {
    position: absolute;
    right: 0;
  }
}

<details
  className="glamor-1"
  open={undefined}
>
  <summary
    className="glamor-2"
  >
    <div
      className="glamor-0"
    >
      <span>
        Problem Solving
      </span>
      <span>
        
      </span>
    </div>
    <svg
      className="glamor-1"
      height={28}
      viewBox="0 0 28 28"
      width={28}
    >
      <g
        fill="none"
        fillRule="evenodd"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
      >
        <polyline
          points="27.155 0 14.224 12.931 0 0"
        />
        <g
          transform="translate(13.577500, 19.465500) scale(1, -1) translate(-13.577500, -19.465500)"
        >
          <polyline
            points="27.155 13 14.224 25.931 0 13"
          />
        </g>
      </g>
    </svg>
  </summary>
  <div
    className="glamor-22"
  >
    <div
      className="glamor-7"
    >
      <label
        className="glamor-6"
        htmlFor="problemSolving"
      >
        <select
          aria-label="Score"
          className="glamor-3"
          id="problemSolving"
          name="problemSolving"
          onChange={[Function]}
          value={undefined}
        >
          <option
            value={null}
          >
            Score
          </option>
          <option
            value={0}
          >
            0 - Bad
          </option>
          <option
            value={1}
          >
            1 - Meh
          </option>
          <option
            value={2}
          >
            2 - Good
          </option>
          <option
            value={3}
          >
            3 - Excellent
          </option>
        </select>
        <div
          className="glamor-5"
        >
          <svg
            className="glamor-1"
            height={12}
            viewBox="0 0 21 12"
            width={21}
          >
            <g
              fill="none"
              fillRule="evenodd"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
            >
              <polyline
                points="19.182 1.067 10.611 10.067 1.182 1.067"
              />
            </g>
          </svg>
        </div>
      </label>
    </div>
    <ol
      className="glamor-11"
    >
      <li
        className="glamor-8"
      >
        Problem Solving Question 1
      </li>
      <li
        className="glamor-8"
      >
        Problem Solving Question 2
      </li>
      <li
        className="glamor-8"
      >
        Problem Solving Question 3
      </li>
    </ol>
    <dl
      className="glamor-21"
    >
      <div
        className="glamor-14"
      >
        <dt
          className="glamor-12"
        >
          2
        </dt>
        <dd
          className="glamor-13"
        >
          Problem Solving Option 3
        </dd>
      </div>
      <div
        className="glamor-14"
      >
        <dt
          className="glamor-12"
        >
          1
        </dt>
        <dd
          className="glamor-13"
        >
          Problem Solving Option 2
        </dd>
      </div>
      <div
        className="glamor-14"
      >
        <dt
          className="glamor-12"
        >
          0
        </dt>
        <dd
          className="glamor-13"
        >
          Problem Solving Option 1
        </dd>
      </div>
    </dl>
  </div>
</details>
`;
